lib/irb.rb

```diff
@@ -979,13 +984,21 @@ def run(conf = IRB.conf)
       end
 
       begin
-        catch(:IRB_EXIT) do
+        forced_exit = false
+
+        forced_exit = catch(:IRB_EXIT) do
           eval_input
         end
       ensure
         trap("SIGINT", prev_trap)
         conf[:AT_EXIT].each{|hook| hook.call}
-        context.io.save_history if save_history
+
+        if forced_exit
+          context.io.save_history if supports_history_saving
+          Kernel.exit!(0)
```

Comment:
I'm changing the method to `Kernel.exit` instead as it achieves the same effect we want but still runs `at_exit` callbacks.

I also don't think it'll be confusing to users as after the PR `exit!` will be listed in the help message as a command.

```suggestion
          Kernel.exit(0)
```